### ISSUE TITLE
rclone missing files from google drive cache:

### ISSUE THREAD

**Issue by zenjabba** [ncw/rclone#1904]
rclone is only seeing 6778 directories in a directory, but I have > 11,000 separate directories in that directory.

When I mount the "mount" directly (without the cache) it finds 13531 directories

What options do you need to be look at?

rclone v1.38-223-g7c972d37β
- os/arch: linux/amd64
- go version: go1.9.2

**Comment by easy90rider** [comment]
I remember seeing something like this. But it was on an old version. 

A new emby/plex library wouldn't scan completely. 

I think  _ls -R_ had better results.

**Comment by remusb** [comment]
Hi

Try to give me some more details. What are the steps performed? Who doesn't see all the files? What operations are being used?

@zenjabba

**Comment by zenjabba** [comment]
ls -l doesn't see all the files when viewed via the cache mount, but does when doing it via the non-cached mount. Simply using `ls -l `displays a different number of files

**Comment by remusb** [comment]
Could the new files be new ones added outside the cache remote? 
The parent directory would cache the listing when you first execute it for `--cache-info-age` time. After that time passes, the cached listing of the parent folder is evicted and a new listing would get the data from the cloud provider.
If you purge the cache and run again, are they still not displayed?

If they don't show up in any case, please provide the log with `-v` and the rclone command and cache config.

Thanks

**Comment by zenjabba** [comment]
No, these files have been in google drive for over a year. This path doesn't have any updates so that isn't it.

I don't know how to purge the cache

```
[zenjabbamisc_cache]
type = cache
remote = zenjabbamisc:/
chunk_size = 10M
info_age = 48h
chunk_age = 12h
warmup_age = 12h
chunk_total_size = 100G
```

**Comment by zenjabba** [comment]
Deleted the cache directory, restarted the mount and they all are back.

**Comment by ElbaRado** [comment]
@zenjabba I'm having the same problem, can you tell me where the directory is located?

**Comment by zenjabba** [comment]
Mine was defined from the command line, and I put it in /tmp

**Comment by jrarseneau** [comment]
I am having the same issue. A good way to find out if you have this is to use the FindMissing.bundle channel for plex: https://github.com/srazer/FindMissing.bundle

For my movies, I have 21 movies that don't appear in the cache mount (but do in the regular encrypted mount) and 1500+ episodes that don't appear.

Most of these were there before using rclone cache, so not sure why they just disappeared.

Let me know how I can help.

**Comment by zenjabba** [comment]
Once I shutdown the cache mount, delete the cache folder and restart it, they appear but then others disappear.

**Comment by zenjabba** [comment]
@remusb if you want real time testing, we can talk on Slack.

**Comment by remusb** [comment]
Can one of you give me the config, mount command and OS to replicate this? I have been unsuccessful so far on both Ubuntu 16.04 and Mac.

@easy90rider @zenjabba @jrarseneau

**Comment by jrarseneau** [comment]
@remusb

Is it possible your library isn’t big enough? I know @zenjabba and I have very big libraries. Mine is over 40TB, his is much bigger I believe.

**Comment by zenjabba** [comment]
Looks sheepishly down at the floor, mine is slightly bigger than 40TB.. I'm happy to give you access to my UAT Plex system for you to connect to my share and test it out. It's sad I have a DEV then UAT then PROD Plex system, but these days you gotta do it properly.

**Comment by easy90rider** [comment]
```
rclone -V
rclone v1.38-223-g7c972d37β
- os/arch: linux/amd64
- go version: go1.9.2

--------------------
[cache]
type = cache
remote = amazon:/
chunk_size = 5M
info_age = 24h
chunk_age = 1m
warmup_age = 3h
chunk_total_size = 5G
--------------------

/usr/bin/rclone mount --read-only --allow-non-empty --cache-db-purge cache: /home/emby/.acd -v --log-file="/home/emby/scripts/logs/rclone-cache-$NOW.txt" &

#Mount encryption over these folders
encfs  /home/emby/.acd /home/emby/acd
encfs /home/emby/.local /home/emby/local

#Use union-fs to merge our remote and local directories
unionfs-fuse -o cow,nonempty /home/emby/local=RW:/home/emby/acd=RO /home/emby/media
exit

```

**Comment by zenjabba** [comment]
rclone v1.38-227-g2cf808c8β
- os/arch: linux/amd64
- go version: go1.9.2

[zenjabbamisc]
type = drive

[zenjabbamisc_cache]
type = cache
remote = zenjabbamisc:
chunk_size = 10M
info_age = 1m
chunk_total_size = 100G

**Comment by remusb** [comment]
Ok, I think I finally got it. Thanks @zenjabba and @jrarseneau for the suggestion on the size. With a larger set of folders I can see the same thing and I think it's related to adding new files in folders that haven't been cached yet or expired. When a list is done on the parent folder it will only see that file and cache won't ask the cloud provider for an actual updated list cause it doesn't think it needs to.

I should have a fix for this

**Comment by jrarseneau** [comment]
Awesome! Looking forward to a new beta with the fix ;)

**Comment by zenjabba** [comment]
Same. What a great christmas present!

**Comment by jrarseneau** [comment]
@remusb new beta soon? ;)

**Comment by remusb** [comment]
In a few, have some testing issues but yes, it's for this issue

**Comment by remusb** [comment]
Here is a new beta that should have fixed this issue. I had to refactor how file info is cached though, there wasn't a fix in that model. I can't say I'm still 100% happy of the caching done now either but as a near term solution it should work out fine.

Some key points:
- cached file info will not be deleted anymore, they simply expire and the next listing will see that and ignore if it's expired
- when a file/folder expires all its parents up to the root will expire too. This is less than ideal but like I said, as a compromise it works. My next milestone is to actually integrate with notifications from the cloud provider if they exist (gdrive has them)

I would strongly recommend to clear the cache before running the next version.
https://beta.rclone.org/v1.38-238-g6f1ae00c/

**Comment by jrarseneau** [comment]
Very nice @remusb 

Just installed it and rebuilding the cache using an `ls -Rl` command. Will let you know how it goes.

I think you're right about the correct approach/next milestone. Using the notification system from the cloud provider (i.e.: Change log for Google Drive) is the better approach.

I mean, there's no point expiring any items in the cache if the change log doesn't send you any changes for those files. I think long term this is more efficient (and the way Plexdrive and other solutions do it as well). 

One thing I've noticed is that when building the cache, everything on the system slows down to an almost halt. If you have bigger libraries with lots of files, this can make for quite a long time your system is bogged down.

Also, due to the nature of Plex scanning/etc, most of the files in the cache in the current model will probably expire at the same time.

**Comment by seuffert** [comment]
@remusb from a quick look at the CleanChunksBySize function it looks like the oldest chunks are deleted first (FIFO?). What do you think about deleting the least recently accessed chunks first instead?

**Comment by remusb** [comment]
Both your comments relate to the feature to keep "hot" data longer in the cache rather than newest data. Which is really what I intended from the beginning with the first model. Seeing as the first model was buggy and causing 503s, that had to change to something more efficient. Sure, it's not the most efficient in terms of Plex scans but at least people won't get 503's anymore.

1.39 will remain as it is I think. I would very much like to start work on a more efficient pattern to combine models (basically what both of you said) after the first release.

**Comment by jrarseneau** [comment]
Been running for about 3 hours now on the new beta and no missing files! Looking good. Will report back tomorrow.

**Comment by floriansofianos** [comment]
What is your rclone version (eg output from `rclone -V`)
v1.38-238-g6f1ae00c

Which OS you are using and how many bits (eg Windows 7, 64 bit)
Windows server 2016 64bit

Which cloud storage system are you using? (eg Google Drive)
Google Drive

The command you were trying to run (eg `rclone copy /tmp remote:tmp`)
rclone mount googleenc: X:

A log from the command with the `-vv` flag (eg output from `rclone -vv copy /tmp remote:tmp`)
When I first run mount, all the files are there and there is no problem.
If I wait about 5 minutes, the issue starts appearing: the sub-folders are all gone.
I also have an error in the logs:
Entry doesn't belong in directory "" (contains subdir) - ignoring
[rclone-log-3.txt](https://github.com/ncw/rclone/files/1569869/rclone-log-3.txt)
I have attached the full log.

Also my config:
[googleenc]
type = crypt
remote = cache:
filename_encryption = standard
password = XXX
password2 = XXX

[cache]
type = cache
remote = gdrive:/Plex-Encrypted/
plex_url = http://127.0.0.1:32400
plex_username = XXX
plex_password = XXX
chunk_size = 20M
info_age = 24h
chunk_total_size = 180G
plex_token = XXX

Thanks for your great work!

**Comment by floriansofianos** [comment]
FYI, I am now testing on a VM running Ubuntu 17.10 with the almost the same configuration: I have just changed chunk_total_size to 10G.
It's all working fine.
So I think the issue is on Windows only.

**Comment by remusb** [comment]
@floriansofianos

You're wrapping drive -> crypt -> cache. It shouldn't cause issues like this but I haven't added tests for this setup cause it has another known issue: it will cause bans very fast from gdrive.

https://github.com/ncw/rclone/blob/master/docs/content/cache.md
It's in the known issues section.

**Comment by floriansofianos** [comment]
@remusb 
I'm doing gdrive > cache > googleenc (drive > cache > crypt), unless there is something I missed.

**Comment by remusb** [comment]
Oh, yeah. You're right. The remote config on the cache section set me off. It has encrypted in the name and I thought it's the encrypted remote

**Comment by remusb** [comment]
@floriansofianos https://beta.rclone.org/v1.38-242-g0d4bff82/ this build should have a fix for the Windows issue you're seeing with `Entry doesn't belong in directory "" (contains subdir)`

**Comment by floriansofianos** [comment]
@remusb Awesome! Thank you very much. I'll test it shortly and let you know how it goes.

**Comment by j0shii** [comment]
The same error keeps appearing..

**Comment by floriansofianos** [comment]
@remusb Yes I have the same error after 5 minutes of use. I cannot attach the logs this time because it's too big but here is what happens before the error:
2017/12/21 02:49:08 DEBUG : : Re-reading directory (5m4.4962215s old)
2017/12/21 02:49:08 DEBUG : : list: warm 225 from cache for: , ts: 2017-12-21 02:44:04.2506533 -0800 PST
2017/12/21 02:49:08 ERROR : TV Shows/xxxxxxxxx: Entry doesn't belong in directory "" (contains subdir) - ignoring
2017/12/21 02:49:08 ERROR : TV Shows/xxxxxxxxx: Entry doesn't belong in directory "" (contains subdir) - ignoring

I hope that helps!

**Comment by remusb** [comment]
@floriansofianos moved the discussion there. We need more investigation and it's clear to me that this issue is different and looks to be happening only on Windows too.

I want to close this one as it was a fundamental change in how caching was performed and let it go for 1.39

**Comment by zenjabba** [comment]
@remusb This shouldn't be closed. I'm still getting missing files when mounting a google drive. If you have another ticket specifically for tracking changes on google drive, I'll start discussions there

**Comment by diggerydoo** [comment]
I am also having identical issues using the latest beta and Windows 10.  Drive > cache > crypt order.  Folders will randomly disappear, but loose files that were not in the folders remain.

**Comment by jkoopmann** [comment]
@ncw I am seeing something similar.

crypt -> cache -> GDRIVE

I mounted something over crypt -> GDRIVE and deleted a file. It kept reappearing in the crypt/cache mounted. I cleared the cache on the directory with rc cache/expire which god rid of the problem. However during Plex runs I still see messages like this:

2018/09/11 01:06:54 ERROR : ndcf2t9p21t408730r32te9c74/7df64qh9kcjl5lcibmoth1oquo/v3mcoa8af43ktmqhqaisovq3qo/0vt4pm36tc4v1amav5uf5id2u8clcppejtat5s4qe8lq105r4bnkii9fv9krjn6lndodqd16jk78m: error refreshing object in : in cache fs Google drive root 'plex': object not found

Do I really need to expire/clear/purge the entire cache now in such scenarios?
